### PR TITLE
rustdoc: Resugar async fn return type in `clean`, not `html`

### LINKED ISSUE #101199: Rustdoc-Json: `async fn foo() -> T` return type is given as `impl Future<Output=T>` not `T`
```rust
pub async fn foo() {}
```

produces:

```json
    "0:4:1568": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:4:1568",
      "inner": {
        "decl": {
          "c_variadic": false,
          "inputs": [],
          "output": {
            "inner": [
              {
                "trait_bound": {
                  "generic_params": [],
                  "modifier": "none",
                  "trait": {
                    "args": {
                      "angle_bracketed": {
                        "args": [],
                        "bindings": [
                          {
                            "args": {"angle_bracketed": {"args": [], "bindings": []}},
                            "binding": {"equality": {"type": {"inner": [], "kind": "tuple"}}},
                            "name": "Output"
                          }
                        ]
                      }
                    },
                    "id": "1:14043:139",
                    "name": "Future"
                  }
                }
              }
            ],
            "kind": "impl_trait"
          }
        },
        "generics": {"params": [], "where_predicates": []},
        "header": {"abi": "Rust", "async": true, "const": false, "unsafe": false}
      },
      "kind": "function",
      "links": {},
      "name": "foo",
      "span": {"begin": [4, 0], "end": [4, 21], "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/async_orphan.rs"},
      "visibility": "public"
    }
```      

But I think it should be.

```json
    "0:4:1568": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:4:1568",
      "inner": {
        "decl": {
          "c_variadic": false,
          "inputs": [],
          "output": {
            "inner": [],
            "kind": "tuple"
          }
        },
        "generics": {"params": [], "where_predicates": []},
        "header": {"abi": "Rust", "async": true, "const": false, "unsafe": false}
      },
      "kind": "function",
      "links": {},
      "name": "foo",
      "span": {"begin": [4, 0], "end": [4, 21], "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/async_orphan.rs"},
      "visibility": "public"
    }
```

For the HTML backend, rustdoc "resugars" the return type to get the `T` from `impl Future<Output=T`>

https://github.com/rust-lang/rust/blob/0631ea5d73f4a3199c776687b12c20c50a91f0d2/src/librustdoc/html/format.rs#L1394-L1401

I think we should move this to a `clean` pass.

cc @CraftSpider @GuillaumeGomez  @Enselic, does this change seem like a good idea.

@rustbot modify labels: +T-rustdoc +A-rustdoc-json

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
This way it also happens for json output.

Fixes #101199

r? @GuillaumeGomez

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks great, thanks! I think it'll even improve performance. Let's check.

@bors try @rust-timer queue

File: src/test/rustdoc-json/fns/async_return.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+use std::future::Future;
+
+// @is "$.index[*][?(@.name=='get_int')].inner.decl.output" '{"inner": "i32", "kind": "primitive"}'
+// @is "$.index[*][?(@.name=='get_int')].inner.header.async" false
+pub fn get_int() -> i32 {
+    42
+}
+
+// @is "$.index[*][?(@.name=='get_int_async')].inner.decl.output" '{"inner": "i32", "kind": "primitive"}'
+// @is "$.index[*][?(@.name=='get_int')].inner.header.async" false
```
Comment (cynecx) [inline src/test/rustdoc-json/fns/async_return.rs:15]:
```suggestion
// @is "$.index[*][?(@.name=='get_int_async')].inner.header.async" true
```

?

Comment (aDotInTheVoid) [reply src/test/rustdoc-json/fns/async_return.rs]: Nice catch, thanks

**Comment by GuillaumeGomez** [comment]
Small improvement in the performance as I thought, well done!

@bors r+ rollup=iffy

**Pull request merged**